### PR TITLE
[Snappi] Validate the PFC frames sent out by the DUT are valid

### PR DESCRIPTION
**Pull request by developfast**
### Description of PR

NOTE: This PR can only be merged after #8660 is merged.

Summary: Validate the PFC frames sent out by the DUT are valid i.e. they meet the 801.1Qbb standards mentioned in this RFC (https://1.ieee802.org/dcb/802-1qbb/) 

For clarity, PFC uses the 802.3 PAUSE frame, and extends it with the following PFC fields:

1) An 8-bit mask that specifies which 802.1p priority levels should be paused.
2) A timer value for each priority specifying how long the traffic for that priority level should be paused.

These two elements are being validated for a sample of PFC frames inspected. 

Fixes (issue)  #8211 

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [X] Test case(new/improvement)

### Back port request
- [ ] 201911
- [X] 202012
- [X] 202205

#### What is the motivation for this PR?
Fill RDMA test gap for validating PFC frames sent out by the DUT

#### How did you do it?
1) Fill up the shared space on the dut to force pfc generation
2) Capture the pfc frame sent by the DUT
3) Validate the packet format

#### How did you verify/test it?
On lab device with a traffic generator

### CODE REVIEW

File: tests/common/snappi/read_pcap.py
Comment (baiwei0427) [inline tests/common/snappi/read_pcap.py:36]: It seems to me that we can define a class for PFC packet, and return the PFC packet object here.

File: tests/common/snappi/read_pcap.py
Comment (baiwei0427) [inline tests/common/snappi/read_pcap.py:39]: pfc_pcap_file

File: tests/common/snappi/read_pcap.py
Comment (baiwei0427) [inline tests/common/snappi/read_pcap.py:44]: frame bytes or pcap file?

File: tests/common/snappi/read_pcap.py
Comment (baiwei0427) [inline tests/common/snappi/read_pcap.py:55]: it seems that we should pass sample_size as an argument to this function and give it a default value

File: tests/common/snappi/read_pcap.py
Comment (baiwei0427) [inline tests/common/snappi/read_pcap.py:78]: Why we need this?

File: tests/common/snappi/read_pcap.py
Comment (baiwei0427) [inline tests/common/snappi/read_pcap.py:111]: I do not quite understand this function. It seems to me that there are many possible combinations since we may enable PFC on any priorities (not only 3 and 4).

File: tests/common/snappi/read_pcap.py
Comment (baiwei0427) [inline tests/common/snappi/read_pcap.py:128]: Since class_pause_time only has two bytes, so it must be in [0, 65535]. But we should check for any priority i, if enable[i] == 0 but pause_duration[i] > 0.

**Review by baiwei0427 — CHANGES_REQUESTED** [review]

File: tests/common/snappi/read_pcap.py
Comment (developfast) [reply tests/common/snappi/read_pcap.py:78]: There's a concern that maybe the switch is not generating enough PFC pause frames within the given time frame

File: tests/common/snappi/read_pcap.py
Comment (developfast) [reply tests/common/snappi/read_pcap.py:111]: The first part of the test traffic enables it on those priorities, hence those are the valid options. We can expand this test later to account for more priorities.

File: tests/common/snappi/read_pcap.py
Comment (developfast) [reply tests/common/snappi/read_pcap.py:128]: Good catch

File: tests/snappi_tests/pfc/test_valid_pfc_frame_with_snappi.py
Comment (baiwei0427) [inline tests/snappi_tests/pfc/test_valid_pfc_frame_with_snappi.py:33]: tgen to snappi

File: tests/snappi_tests/pfc/test_valid_pfc_frame_with_snappi.py
Comment (baiwei0427) [inline tests/snappi_tests/pfc/test_valid_pfc_frame_with_snappi.py:36]: How can you create congestion with 1 sender and 1 receiver? I guess you create congestion by letting 1 send data traffic to 2, and 2 send PFC pause frames to DUT.

File: tests/snappi_tests/pfc/test_valid_pfc_frame_with_snappi.py
Comment (baiwei0427) [inline tests/snappi_tests/pfc/test_valid_pfc_frame_with_snappi.py:39]: In addition to these comments, we should have a test case document.

**Review by baiwei0427 — CHANGES_REQUESTED** [review]

File: tests/snappi_tests/pfc/test_valid_pfc_frame_with_snappi.py
Comment (developfast) [reply tests/snappi_tests/pfc/test_valid_pfc_frame_with_snappi.py:36]: Yes that's right

File: tests/common/snappi_tests/pfc_packet.py
Comment (baiwei0427) [inline tests/common/snappi_tests/pfc_packet.py:124]: It seems that we should add a function called is_valid to return is_valid_frame.

File: tests/snappi_tests/pfc/test_valid_pfc_frame_with_snappi.py
Comment (baiwei0427) [inline tests/snappi_tests/pfc/test_valid_pfc_frame_with_snappi.py:36]: This is done by letting 1 send data traffic to 2

**Review by baiwei0427 — CHANGES_REQUESTED** [review]

File: tests/common/snappi_tests/pfc_packet.py
Comment (developfast) [reply tests/common/snappi_tests/pfc_packet.py:124]: done

**Review by baiwei0427 — APPROVED** [review]

**Review by bingwang-ms — APPROVED** [review]

**Comment by developfast** [comment]
Hold off on merging to other branches until #9443 and #9444 is merged. This test relies on the PFC directory being refactored, and as such will create a merge conflict right now.

**Comment by wangxin** [comment]
@developfast This is to add new tests. Do we really need it in 202012 branch?

**Comment by developfast** [comment]
Yes @wangxin, we currently have snappi nightly pipelines running 202012 branch.

**Comment by yxieca** [comment]
202012 branch cherry-pick is pending after #8661

**Pull request merged**
